contracts/src/CrossChainPingPong.sol

```diff
@@ -0,0 +1,96 @@
+
+error InvalidCrossDomainSender();
+
+struct PingPongBall {
+    uint256 rallyCount;
+    uint256 lastHitterChainId;
+    address lastHitterAddress;
+}
+
+contract CrossChainPingPong {
+    event BallSent(uint256 indexed toChainId, PingPongBall ball);
+
+    event BallReceived(uint256 indexed fromChainId, PingPongBall ball);
+
+    address internal constant MESSENGER = Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER;
+
+    mapping(uint256 => bool) internal allowedChainIds;
+
+    uint256 internal serverChainId;
+    bool internal serverAlreadyServed;
+
+    PingPongBall internal receivedBall;
+
+    constructor(uint256[] memory _allowedChainIds, uint256 _serverChainId) {
+        for (uint256 i = 0; i < _allowedChainIds.length; i++) {
+            allowedChainIds[_allowedChainIds[i]] = true;
+        }
+
+        require(allowedChainIds[_serverChainId], "Invalid first server chain ID");
+
+        serverChainId = _serverChainId;
+    }
+
+    function serveBall(uint256 _toChainId) external {
+        require(serverChainId == block.chainid, "Cannot serve ball from this chain");
+        require(serverAlreadyServed == false, "Ball already served");
+        require(_isValidDestinationChain(_toChainId), "Invalid destination chain ID");
+
+        serverAlreadyServed = true;
+
+        PingPongBall memory _newBall = PingPongBall(1, block.chainid, msg.sender);
+
+        _sendBallMessage(_newBall, _toChainId);
+
+        emit BallSent(_toChainId, _newBall);
+    }
+
+    function sendBall(uint256 _toChainId) public {
+        require(_isBallOnThisChain(), "Ball is not on this chain");
+        require(_isValidDestinationChain(_toChainId), "Invalid destination chain ID");
+
+        PingPongBall memory _newBall = PingPongBall(receivedBall.rallyCount + 1, block.chainid, msg.sender);
+
+        delete receivedBall;
+
+        _sendBallMessage(_newBall, _toChainId);
+
+        emit BallSent(_toChainId, _newBall);
+    }
+
+    function receiveBall(PingPongBall memory _ball) external {
+        if (msg.sender != MESSENGER) {
+            revert CallerNotL2ToL2CrossDomainMessenger();
+        }
+
+        if (IL2ToL2CrossDomainMessenger(MESSENGER).crossDomainMessageSender() != address(this)) {
+            revert InvalidCrossDomainSender();
+        }
+
+        receivedBall.lastHitterAddress = _ball.lastHitterAddress;
+        receivedBall.lastHitterChainId = _ball.lastHitterChainId;
+        receivedBall.rallyCount = _ball.rallyCount;
+
+        emit BallReceived(IL2ToL2CrossDomainMessenger(MESSENGER).crossDomainMessageSource(), _ball);
+    }
+
+    function _sendBallMessage(PingPongBall memory _ball, uint256 _toChainId) internal {
+        bytes memory _message = abi.encodeCall(this.receiveBall, (_ball));
+        IL2ToL2CrossDomainMessenger(MESSENGER).sendMessage(_toChainId, address(this), _message);
+    }
+
+    function _isBallOnThisChain() internal view returns (bool) {
+        return receivedBall.lastHitterAddress != address(0);
```

Comment:
if checking for the ball on the chain, why does this check the lastHitterAddress? Is it just meant to check that receivedBall is defined so this check could really be for any of the fields on `receivedBall`? nit but it might be good to add a comment for why this check is done this way, since I am not sure how intuitive that is from just reading this one line of code and not having an understanding of the entire contract.

Reply:
yeah I should've added more comments - this is just a hacky way of checking the receivedBall is defined - will add a comment!